compiler/passes/ResolveScope.cpp

```diff
@@ -529,72 +528,147 @@ Symbol* ResolveScope::followImportUseChains(UnresolvedSymExpr* expr) const {
+          retval = sym;
+          break;
+        } else if (mod->defPoint->parentSymbol != theProgram) {
           // if we're not in the root module scope, this is an improper match
-          USR_WARN(expr, "currently unable to import nested modules without"
-                   " specifying the full path to the module");
-          USR_PRINT(mod, "skipping module defined here");
-          foundBadCloser = true;
+          badCloserModule = mod;
         } else {
           // if we are in the root module scope, then it is a proper match.
           retval = sym;
           break;
         }
       } else if (sym != NULL) {
-        retval = sym;
-        break;
+        // found something other than a module
+        USR_FATAL(expr, "import must name a module ('%s' not a module)",
+                  sym->name);
       }
 
       // otherwise follow uses/imports only (breadth first)
       //    look for the symbols defined in the used
       //    or if the module being imported matches
       //    and if not, keep following the public use chains
-      retval = followImportUseChains(uSym);
+      retval = followImportUseChains(name);
+
+      if (relativeScope != NULL)
+        break; // only consider the one scope when doing relative imports
     }
 
-    if (retval == NULL && foundBadCloser) {
-      USR_FATAL(expr, "No better match found");
+    if (retval == NULL && badCloserModule != NULL) {
+      USR_FATAL_CONT(expr, "Cannot import module '%s'", name);
+      USR_PRINT(badCloserModule, "a module named '%s' is defined here", name);
+      USR_PRINT(expr, "full path or explicit relative import required");
+      USR_PRINT(expr, "please specify the full path to the module");
+      USR_PRINT(expr, "or use a relative import e.g. 'import this.M' or 'import super.M'");
+      USR_STOP();
     }
-  } else if (CallExpr* call = toCallExpr(expr)) {
-    if (call->isNamedAstr(astrSdot) == true) {
-      Symbol* outer = lookupForImport(call->get(1));
-      if (ModuleSymbol* outerMod = toModuleSymbol(outer)) {
-        if (SymExpr* rhs = toSymExpr(call->get(2))) {
-          const char* rhsName = NULL;
+  }
 
-          if (get_string(rhs, &rhsName) == true) {
-            ResolveScope* scope = getScopeFor(outerMod->block);
-            if (Symbol* symbol = scope->getField(rhsName)) {
-              retval = symbol;
+  if (retval == NULL)
+    USR_FATAL(expr, "Cannot find symbol '%s'", name);
 
-            } else {
-              USR_FATAL(call, "Cannot find symbol '%s'", rhsName);
-            }
-          } else {
-            INT_FATAL(call, "Bad qualified name");
-          }
+  // Process further portions of import starting from call
+  while (call != NULL) {
+    INT_ASSERT(call->isNamedAstr(astrSdot));
+    if (!isModuleSymbol(retval))
+      USR_FATAL(call, "cannot make nested import from non-module '%s'",
+                 retval->name);
+
+    Symbol* outer = retval;
+    ModuleSymbol* outerMod = toModuleSymbol(outer);
```

Comment:
What about `super.M.N.O`?  I would argue that this triggers a different scenario than `super.M.N`, and so should be checked.  I don't think `super.M.N.O.P` will be significantly different than `super.M.N.O` - if the latter works, then I think it is less likely the former will be broken.

In a little more detail - I know that deep calls at this point are represented as `call(. call(. call(. super "M") "N") "O")`, which means that `get(1)` for the outermost call is `call(. call (. super "M") "N")`, so it seems possible that what you have implemented will accidentally look for `super.M.O` instead of `super.M.N.O`

Reply:
The firstImportedModuleName should return the call `call(. super "M") "N")` in that case and then the loop traverses up parentExprs to process the remainder. Anyway I'll add a test for this case.

Reply:
I justed pushed the commit adding a test import-relative-super3.chpl that does this